2022/day7/day7.hs

```diff
@@ -0,0 +1,132 @@
+  putStrLn ("Total size of directories within limit: " ++ show sumWithinLimit ++ " [PART ONE ANSWER]")
+
+  -- Part Two: Find directories with accumulated sizes above the space needed to free up.
+  --           Then identify the smallest directory that can be deleted to free up enough space.
+  --           To prepare, we first need to calculate how much disk space is used and how
+  --           much space needs to be freed up. We do that by finding the root directory's size.
+  let diskCapacity = 70000000
+  let spaceNeeded  = 30000000
+
+  let totalSizeUsed = maybe 0 snd (find (\(path, size) -> length path == 0) dirEntries)
+  let (freeSpace, spaceToFree) = (diskCapacity - totalSizeUsed, spaceNeeded - freeSpace)
+  let bigEnough = (\(_, size) -> size >= spaceToFree)
+  let indicateBigEnough = (\entry -> if bigEnough entry then Just " ** BIG ENOUGH **" else Nothing)
+
+  let (dirToDelete, spaceFreed) = findSmallest (filter bigEnough dirEntries)
+
+  putStrLn ("")
+  putStrLn ("~~~~~~~~~~~~~~~~~~ PART TWO ~~~~~~~~~~~~~~~~~~")
+  putStrLn (formatEntries indicateBigEnough (sortByPath dirEntries))
+  putStrLn ("")
+  putStrLn ("Total Disk Capacity : " ++ show diskCapacity)
+  putStrLn ("Space Needed        : " ++ show spaceNeeded)
+  putStrLn ("Total Space Used    : " ++ show totalSizeUsed)
+  putStrLn ("Total Space Free    : " ++ show freeSpace)
+  putStrLn ("Space to Free       : " ++ show spaceToFree)
+  putStrLn ("")
+  putStrLn ("Directory to delete : " ++ formatPath dirToDelete)
+  putStrLn ("Space to be Freed   : " ++ show spaceFreed ++ " [PART TWO ANSWER]")
+
+
+findSmallest :: [(path, Int)] -> (path, Int)
+findSmallest = minimumBy (compare `on` snd)
+
+accumulateSizes :: [(path, Int)] -> (Int, [(path, Int)])
+accumulateSizes = mapAccumL (\acc (path, size) -> (acc + size, (path, size))) 0
+
+isDescendent :: ([String], Int) -> Bool
+isDescendent entry = length (fst entry) > 0
+
+entriesWithAncestors :: [([String], Int)] -> [([String], Int)]
+entriesWithAncestors = filter isDescendent
+
+parentEntry :: ([String], Int) -> ([String], Int)
+parentEntry (path:parent, size) = (parent, size)
+
+parentEntries :: [([String], Int)] -> [([String], Int)]
+parentEntries fileEntries = map parentEntry (entriesWithAncestors fileEntries)
+
+flattenAncestors :: [([String], Int)] -> [([String], Int)]
+flattenAncestors [] = []
+flattenAncestors fileList = toList(fromListWith (+) (fileList ++ flattenAncestors (parentEntries fileList)))
+
+processLog :: String -> [([String], Int)]
+processLog commandLog = parseCommandLines (DirectoryListing [] [], lines commandLog)
+
+parseCommandLines :: (DirectoryListing [String] [String], [String]) -> [([String], Int)]
+parseCommandLines (listing, []) = files listing
+parseCommandLines (listing, ('$':' ':command) : logLines) = parseCommandLines (parseCommand (listing, command, logLines))
+
+parseCommand :: (DirectoryListing [String] [String], [Char], [String]) -> (DirectoryListing [String] [String], [String])
```

Comment:
~~There are a lot of primitive type names in this declaration (and others). Does Haskell have an easy way to create some kind of type alias? Could this be redefined in terms of a `Command` and `Argument` instead of `String` and `String`?~~

Nevermind. Now that I think about it, a tool probably/maybe wrote the type signature, and it's about parsing; it _has_ to deal with a bunch of primitives at some point.

Reply:
The signature was written by hand. But I'm going through and simplifying many of the signatures.